### ISSUE TITLE
Build view based List.

### ISSUE THREAD

**Issue by typesupply** [robotools/vanilla#138]
List was written in 2005. A lot has changed in NSTableView and the best practices for working with NSTableView in the, OMG, last 16 years. It is showing its age, has hit some limits and relies on many things that Apple now labels as "legacy." I've started it and created a new branch for it. This won't be backwards compatible with List*, so it will be called List2.

*No, really, we shouldn't try. List is **extremely** complicated and has so many alternate behavior allowances that the code will be unwieldy. I tried to build all of the new stuff on top of the existing List and it made my head hurt so bad that I literally had to take a week off from thinking about it. We'll leave List in place and unchanged for backwards compatibility.

---

To Do:
- [X] Switch to view base instead of cell base.
- [X] Use vanilla for building prototype views.
- [X] Don't use Cocoa Bindings.
- [x] Enable sorting.
- [x] Drag and drop.
- [x] selectionCallback
- [x] doubleClickCallback
- [x] editCallback
- [ ] menuCallback
- [ ] ~enableDelete~ deleteKeyCallback
- [x] [enableTypingSensitivity](https://developer.apple.com/documentation/appkit/nstableview/1526084-allowstypeselect?language=objc)
- [x] allowsMultipleSelection
- [x] allowsEmptySelection
- [x] draw lines
- [x] getNSScrollView
- [x] getNSTableView
- [x] overridable classes
- [x] enable
- [x] set
- [x] get
- [x] getArrangedItems
- [ ] get and getArrangedItems as iterators
- [ ] getEditedColumnAndRow
- [x] getSelection
- [x] setSelection
- [X] getSelectedItems
- [X] setSelectedItems
- [x] scrollToSelection
- [ ] setMenu
- [x] [column reordering](https://developer.apple.com/documentation/appkit/nstableview/1530207-allowscolumnreordering?language=objc)
- [x] [column resizing masks](https://developer.apple.com/documentation/appkit/nstablecolumn/1529591-resizingmask?language=objc)
- [x] [column autosaving](https://developer.apple.com/documentation/appkit/nstableview/1525596-autosavetablecolumns?language=objc)
- [x] [group rows](https://developer.apple.com/documentation/appkit/nstableviewdelegate/1526676-tableview?language=objc)
- [x] global allowsSorting
- [x] CheckBoxListCell
- [x] SliderListCell
- [x] PopUpButtonListCell
- [x] ImageListCell
- [x] SegmentedButtonListCell
- [x] ColorWellListCell

---

Thinking:

- Allow a dict as the items source instead of list.
- **Done** Sorting needs to be figured out before getting too deep. I ma have to wrap objects with an NSObject for that [NSSortDecriptor](https://developer.apple.com/documentation/foundation/nssortdescriptor?language=objc) can work.
- Drag and drop is too complicated.
    - https://www.natethompson.io/2019/03/23/nstableview-drag-and-drop.html
    - https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DragandDrop/DragandDrop.html

---

Tests:

- [X] Selection Indexes
- [x] Cell types
- [x] Sorting

---

Bugs:

- [ ] Preserve selected items after sort, reload, etc.
- [ ] Text is not properly aligned vertically.
- [ ] drag and drop reordering algorithm

**Comment by typesupply** [comment]
cc @typemytype @LettError @justvanrossum @benkiel in case you have any thoughts or requests.

**Comment by justvanrossum** [comment]
Agreed with a clean start as `List2`.

**Comment by benkiel** [comment]
Seconded

**Comment by typesupply** [comment]
@justvanrossum

Theory question: I'm implementing the "let the user sort the table by clicking the column header" thing this time around. Sorting is handled in the data source by the [`tableView_sortDescriptorsDidChange_`](https://developer.apple.com/documentation/appkit/nstableviewdatasource/1532935-tableview?language=objc) method. This, and the [corresponding documentation on sorting](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/TableView/SortingTableViews/SortingTableViews.html#//apple_ref/doc/uid/10000026i-CH10-SW1), seem to imply that the order of the items within the data source should be changed based on the given sort descriptors. Am I reading this correctly? This is an abstraction of how I have it implemented:

```python

class List2DataSource(NSObject):

    def initWithItems_(self, items):
        self._items = items
        # rest of init

    def tableView_sortDescriptorsDidChange_(tableView, sortDescriptors):
        items = self._items
        sortDescriptors = tableView. sortDescriptors()
        for sortDescriptor in sortDescriptors:
            items = sorted(items, ...)
        self._items = list(items)
        tableView.reloadData()
```

It seems *really unexpected* to sort the actual data in the data source. If two NSTableViews use the same data source things will go awry. Maybe I'm supposed to keep both the original order and the sorted order? I'm now remembering why I didn't implement this in 2005...

**Comment by justvanrossum** [comment]
List is a view on the data, and sorting should alter the view, not the data. The data source is the abstraction between List and actual data. It may be better to maintain a list of indices than a sorted copy of the underlying data.

'List.__getitem__` etc. should probably reflect the sorted order? Selection indices as well? 

You should be able to do all sorting with a single `sorted()` call, capturing all sort descriptors in a single key func.

**Comment by justvanrossum** [comment]
If the data source is responsible for sorting, then I don't think two List views should share data sources. Syncing two Lists looking at the same data would then be up to the client.

**Comment by typesupply** [comment]
> List is a view on the data, and sorting should alter the view, not the data. The data source is the abstraction between List and actual data. It may be better to maintain a list of indices than a sorted copy of the underlying data.

That's how it is in my mind, too. I just reread the programming guide again and noticed this buried at the end:

>> The implementation shown in Listing 7-1 sorts the current model content using the table view sort descriptors, but it also loses the initial order of the data. If the initial order is important to your app, you’ll want to take steps to cache the original order so that it can be restored

So, yeah, I need to keep the original order. No big deal since the data source will only be used for one List2.

> List.__getitem__ etc. should probably reflect the sorted order? Selection indices as well?

This is what is making my head hurt. Say a script creates a List2, `self.w.l`, with this order:

```
B
C
A
```

The user clicks the header to sort. They now see:

```
A
B
C
```

What should the script get when asking for `self.w.l[0]`, A or B? Is the script referencing 0 of the underlying data or the presented data?

> You should be able to do all sorting with a single sorted() call, capturing all sort descriptors in a single key func.

Hm. I was following the [complex sorting model presented here.](https://docs.python.org/3/howto/sorting.html#sort-stability-and-complex-sorts) Would it be better to create a function on the fly to use as the key?

**Comment by justvanrossum** [comment]
Multisort, aha, that is a good example. Perhaps multiple passes is indeed better, esp. if the reverse flag differs per column.

Script looking at List: I'm thinking the List view should give the presented order. If you need the original order: look at the underlying data. I think this gives us the best of both worlds.

**Comment by typesupply** [comment]
> Script looking at List: I'm thinking the List view should give the presented order. If you need the original order: look at the underlying data. I think this gives us the best of both worlds.

Okay. I'll say something about this in the documentation. Does this make sense?

_The presented data may have a different order than the original data. It should not be expected that the item at the index in the presented list is the same as the item at the same index in the original order. For example, the `listItem` and `dataItem` in `buttonCallback` may not be the same item._

```python
class MyWindow:

    def __init__(self):
        ...
        self.data = ["B", "C", "A"]
        self.w.l = vanilla.List2(
            ...,
            items=self.data,
            selectionCallback=self.lscb
            ...
        )
        ...

    def buttonCallback(self, sender):
        index = self.w.l.getSelection()[0]
        listItem = self.w.l[index]
        dataItem = self.data[index]
```

**Comment by justvanrossum** [comment]
Btw. I'm not 100% convinced of that behavior, but it seems to make some intuitive sense.

The `__getitem__` etc. part of the `List` api is a little funny anyway. I think it's mostly a convenience for simple cases: you throw some data at `List` and once the user is done you get the data back out from `List` via `__getitem__`.

But: what does this all mean in the context of manual reordering with drag and drop? Does that even go together at all? If sorting is active, you can't meaningfully drop at a specific location. But what if you want both?

Perhaps two modes need to be supported:

1. Order is "virtual", a view on the original data. Item insertion will insert it at the sorted location (Finder list view).
2. Order is "concrete": the data needs to be actually sortable by the user. You can drop an item at a specific location. Sorting is an action on the data itself, not just the view. (For example, sorting the axes in a designspace file by tag, name, or manually with drag and drop.)

Just thinking out loud here. Am I making this too complicated?

**Comment by justvanrossum** [comment]
View filtering is also very useful: show only rows that match xyz in column 5. I don't know if List should support this fully out of the box, but perhaps we can think about this. How can a user implement such behavior, and how could List help? Perhaps this is out of scope.

**Comment by typesupply** [comment]
Thinking about the big picture... This complexity is arising because of one potential event: the user clicks a header to reorder the virtual data. I can't think of an instance in OS X in which this action has transformed the concrete data. If the order is meaningful, the table doesn't allow sorting via the column header. For example, an interface that allows changing font.glyphOrder by clicking the header seems like bad UI. Allowing sorting is optional and off by default.

I'm starting to feel that all programmatic interaction should *always* work with concrete order.

**Comment by typesupply** [comment]
> View filtering is also very useful: show only rows that match xyz in column 5. I don't know if List should support this fully out of the box, but perhaps we can think about this. How can a user implement such behavior, and how could List help? Perhaps this is out of scope.

I forgot to mention: I think this is in scope and it's something I was thinking about a couple of weeks ago. I don't have any API ideas yet, but I'd like for filtering to be available somehow.

**Comment by LettError** [comment]
In some tools it is expected that the order displayed in the list is also the order in which items are written to file. Is that what you mean?

**Comment by typesupply** [comment]
> In some tools it is expected that the order displayed in the list is also the order in which items are written to file. Is that what you mean?

Yes. That seems like a case when sorting should not be enabled.

**Comment by justvanrossum** [comment]
Ok, fair enough. Let me see if I can recap:

1. If table header sorting is enabled at all, then it is a sorted *view*, which does not change the underlying data.
1.1 Drag and drop does not work within the list
1.2 Dropping an item will insert in sort order in the view.
1.3 But where will it be inserted in the underlying data? Or with what kind of API will the underlying data be edited in this case?
1.4 Can the underlying data be a dictionary? _Should_ the underlying data be a dictionary?

2. With table header sorting disabled, an explicit order can be maintained, and drag & drop within the list will reorder the underlying data. An application could provide a menu to perform sorting, but header sorting is not the right UI in this case.

Is it perhaps a dict vs. list kindof dichotomy?

**Comment by typesupply** [comment]
> Ok, fair enough. Let me see if I can recap:

> 1. If table header sorting is enabled at all, then it is a sorted view, which does not change the underlying data.
> 1.1 Drag and drop does not work within the list
> 1.2 Dropping an item will insert in sort order in the view.

Yes to both. This is the behavior in Finder.

> 1.3 But where will it be inserted in the underlying data? Or with what kind of API will the underlying data be edited in this case?

In this case, a UI wouldn't allow a specific row insertion point. [See the second paragraph in "Discussion" here.](https://developer.apple.com/documentation/appkit/nstableview/1535123-setdroprow?language=objc)

> 1.4 Can the underlying data be a dictionary? Should the underlying data be a dictionary?

This is one of those times when I have something working but don't have the words to explain it. Bear with me...

- We have a list of items in the data source.
- Each row corresponds to one item from the list of items.

An item is a Python object. I'm not using Cocoa Bindings at all, so there is no conversion to `NSWhatever`.

- Columns are defined with dict based descriptions during `__init__`.
- Each column has a unique identifier.
- Each column corresponds is mapped to a value of the item.

This is where things are different from before. You can now specify how the value is get from set to the item. The options are property, method name or a function. If none of these is given, the value will be retrieved with `item[columnIdentifier]` (thus, like a dict). If a simple type is given tio display in the list, behind the scenes I convert it to a dict just like in the current List implementation. This allows us to use dicts or custom objects and define get/set with flexibility.

> 2. With table header sorting disabled, an explicit order can be maintained, and drag & drop within the list will reorder the underlying data. An application could provide a menu to perform sorting, but header sorting is not the right UI in this case.

Yes.

**Comment by justvanrossum** [comment]
1.3 I get the drop behavior. But if the underlying data is list-like, where will it be inserted?

1.4 I get that row objects can be anything, and your proposal makes sense. What I meant, though, is what if the entire underlying data object were dict-like instead of list-like? Eg. can we make a List view on a glyphset directly?

**Comment by typesupply** [comment]
> 1.3 I get the drop behavior. But if the underlying data is list-like, where will it be inserted?

I see your point. Maybe this can be solved in the drag and drop API. I've had a few requests to make that easier than it currently is, so it may get a full rewrite.

> 1.4 I get that row objects can be anything, and your proposal makes sense. What I meant, though, is what if the entire underlying data object were dict-like instead of list-like? Eg. can we make a List view on a glyphset directly?

Ahhhhh. That's a good question. I'll think about it while I'm building.

I've been wondering about doing away with indexes entirely in favor of items. For example, `myList.setSelection([thisItem, thatItem])`. I _always_ have a redundant dance when `selectionCallback` is called:

```python
indexes = sender.getSelection()
if not indexes:
    return
items = sender[i for i in indexes]
# now finally do something with the indexes
```

This would be nicer:

```python
items = sender.getSelection()
```

That's a pretty big difference from vanilla.List so I'm hesitant.

**Comment by benkiel** [comment]
No that’s super nice and does away with a lot of unnecessary code

**Comment by justvanrossum** [comment]
Both are necessary. Perhaps split them like so:
- `getSelectedIndices()`
- `getSelectedItems()`

**Comment by typemytype** [comment]
a bit late tot the game, some thoughts:

1. `List2` is perfect, it allows vanilla to make changes which are not compatible with the current `vanilla.List`, for the better.

2. dont provide support for `__getitem__` aka `list[0]` and work with the items and the displayed items: 
- `list.items()` an iterator over the input data, raw data 
- `list.arrangedItems()` an iterator over the items as how they are displayed (different sorting, applied search queries where less items are shown) 

3. I would go for `getSelectedItems()` and `setSelectedItems()` and dont provide support for selection indexes anymore (indexes are a useless when working with queries) . 

4. Drop and insert into an sorted list: add the dropped item(s) in the raw data before (or after) the visual item (in case of a dict this is not important)

**Comment by typesupply** [comment]
> applied search queries where less items are shown

Is there a Cocoa mechanism for filtering? I don't recall seeing one, but I haven't done a thorough search.

**Comment by typemytype** [comment]
cocoa has [NSPredicate](https://developer.apple.com/documentation/foundation/nspredicate?language=objc) objects for queries

see the [Mechanic toolbar search queries on a array controller](https://github.com/robofont-mechanic/mechanic-2/blob/f110f60abfbf6de07f11a64284a043a2598a553c/Mechanic2.roboFontExt/lib/mechanic2/ui/controller.py#L536-L537)

**Comment by justvanrossum** [comment]
> I would go for getSelectedItems() and setSelectedItems() and dont provide support for selection indexes anymore (indexes are a useless when working with queries) .

Hard disagree. Internally the List will still need indices, and finding the index based on the item does not scale to long lists. This is a performance trap waiting to happen.

**Comment by justvanrossum** [comment]
It also doesn't work if an item occurs in the list multiple times. `getSelectedItems()` is useful, `setSelectedItems()` is a mistake.

**Comment by justvanrossum** [comment]
> cocoa has NSPredicate objects for queries
> see the Mechanic toolbar search queries on a array controller

Not a big fan of these kinds of Cocoa mechanisms. Works well if you work with pure Cocoa objects (NSArray containing NSObjects, using an NSArrayController, etc.), not so much if you want to filter a Python list: it will back and forth through the objc bridge many times, and will be slow with many objects. I think we should focus making things work easy and efficient with pure Python data structures.

**Comment by typesupply** [comment]
>  Internally the List will still need indices, and finding the index based on the item does not scale to long lists. This is a performance trap waiting to happen.

> It also doesn't work if an item occurs in the list multiple times.

These are concerns that have been looming over me. I'm facing it in three different places:

- `setSelectedItems`
- `getSelectedIndexes`
- `setSelectedIndexes`

Within the data source/delegate object I maintain two lists:

- `items` The original order and full list.
- `arrangedItems` The sorted order and, possibly, filtered list.

`items` is never touched apart from extracting items for arrangement. `arrangedItems` is always used for populating the data. So, NSTableView doesn't know anything about the original order, only the arranged order. When I get the selection, I do this:

```python
    def setSelectedItems(self, items):
        arrangedItems = self._dataSourceAndDelegate.arrangedItems()
        indexes = [
            arrangedItems.index(item)
            for item in items
            if item in arrangedItems
        ]
        indexes = makeIndexSet(indexes)
        self._tableView.selectRowIndexes_byExtendingSelection_(indexes, False)

    def getSelectedIndexes(self):
        selectedItems = self.getSelectedItems()
        items = self.get()
        indexes = [
            items.index(item)
            for item in selectedItems
        ]
        return indexes

    def setSelectedIndexes(self, indexes):
        items = self.get()
        arrangedItems = self.getArrangedItems()
        arrangedIndexes = []
        selectedItems = [items[i] for i in indexes]
        self.setSelectedItems()
```

As Just said, there are obviously problems here:

1. `items.index(item)` relies on object comparison. If two different objects are "the same" this will return the first match which may not be the correct one. Scenario that will almost certainly happen: two dicts with identical contents
2. This could scale poorly. If the comparison does some kind of complex data comparison rather than simple id comparison, this could cause a significant performance hit even with a small number of items. Scenario from the past: RoboFab's glyph objects compared their contents for equivalency and it was very expensive.

I haven't dug too deeply into this yet, but these are the 1% thought through ideas that have popped into my mind when I have thought about it.

- maintain a `{ordered index : arranged index}` and vice versa mappings for translations in `getSelectedIndexes` and `setSelectedIndexes`.
- document that `setSelectedItems` is a potential performance problem under certain conditions.

Does anyone have a better idea? My head has mostly been in dealing with visual issues so I haven't given this a ton of thought.

**Comment by justvanrossum** [comment]
Please do not add `setSelectedItems()`. It's an idiotic, not scaling, not working thing. Kill it with fire.

Please to not implement `getSelectedIndexes()` in terms of `getSelectedItems(). Doing it that way has the same problem with duplicate items. The table view itself works with indices, so let's use those.

I would recommend against mainting an "arranged items" list. (Unless you're using Cocoa machinery for this? Is NSArrayController being used here? Isn't that part of Cocoa bindings? I'm probably mixing things up.

I would keep a single list of indices into the original data (keys if the data is a dict). This list corresponds to what's visible and how it's visibly sorted, and references the original data by index (or key, if the data is a dict). A reverse mapping can be made quickly from this, perhaps can be maintained to rebuild only once the main index list changes (on sort/filter).

Whenever you find yourself using `lst.index(...)` or `if item in lst` I can bet you something is wrong or won't scale. In this section you do both: `if item in arrangedItems` loops over `arrangedItems`, and then `arrangedItems.index(item)` does it again:

```python
        indexes = [
            arrangedItems.index(item)
            for item in items
            if item in arrangedItems
        ]
```

**Comment by justvanrossum** [comment]
(Please completely disregard everything I said if NSArrayController is involved -- I have zero experience with that.)

**Comment by typesupply** [comment]
> Is NSArrayController being used here

No. I'm trying to keep this a close to Python as I can. The Cocoa Bindings stuff makes it too complicated for what this is for.

> Whenever you find yourself using lst.index(...) or if item in lst I can bet you something is wrong or won't scale.

Yeah. This is just my quick implementation while I try to figure out how the bigger pieces fit together.

**Comment by typesupply** [comment]
> Please do not add setSelectedItems(). It's an idiotic, not scaling, not working thing. Kill it with fire.

These are the things that I'm trying to balance in my head, so I have mixed feelings on this. It does present an obvious scale problem. On the other hand, how often is vanilla going to be used to display 100,000 items vs how often it will be used to display 100 items? Is the potential for performance problems more of a concern than the convenience of `self.w.glyphList.setSelectedItems([font["A", font["Z"])`? The programmer in me is saying "No, don't make this possible." The scripter in me is saying, "Ugh. I don't want to have to deal with indexes if I don't have to."

**Comment by justvanrossum** [comment]
I find the "duplicate item" issue worse than the scaling issue. But if people agree that it's a useful convenience that is worth a compromise, I'll gladly back off, as long as the scaling and duplicate item problems stay isolated to `setSelectedItems()`. (I think it's a bit like asking a dict for the key belonging to a value: Python's dict doesn't offer any convenience to achieve that, because it's fundamentally the wrong thing to ask of a dict.)

Your example highlights the usefulness of a dict-based List view. I think this:

`self.w.glyphList.setSelectedItems([font["A"], font["Z"]])`

could then be nicely written like this:

`self.w.glyphList.setSelection(["A", "Z"])`

**Comment by typesupply** [comment]
> Your example highlights the usefulness of a dict-based List view.

I've been thinking about this. Would you allow both dict and list as bases?

**Comment by justvanrossum** [comment]
> Would you allow both dict and list as bases?

Yes. Both have their place.

When thinking about the dict-ike data possibility, `getSelection()` would return a list of keys, but then for list-like data it should perhaps return indices into the original data after all, not into the sorted/filtered/arranged view. Going back and forth about this.

Thinking about my earlier question about dropping something on a sorted list, with list-like data: "where to insert it in the underlying data?" I think this is the responsibility of the drop handler anyway, so nothing for List2 to worry about.

**Comment by typesupply** [comment]
This should be possible. I'll get the list based version working and then handle a dict based version.

**Comment by typesupply** [comment]
Here's another place where duplicate items, or items that compare the same, will be a problem:

```python
    def _updateArrangedItems(self):
        tableView = self._tableView
        sortDescriptors = tableView.sortDescriptors()
        items = self._items
        for sortDescriptor in reversed(sortDescriptors):
            identifier = sortDescriptor.key()
            ascending = sortDescriptor.ascending()
            reverse = not ascending
            getters = self._valueGetters.get(identifier, {})
            property = getters.get("property")
            method = getters.get("method")
            function = getters.get("function")
            if property is not None:
                key = operator.attrgetter(property)
            elif method is not None:
                key = operator.methodcaller(method)
            elif function is not None:
                key = function
            else:
                key = operator.itemgetter(identifier)
            items = sorted(
                items,
                key=key,
                reverse=reverse
            )
        self._arrangedItems = list(items)
        tableView.reloadData()
```

I lose the ability to differentiate these items when `sorted` is called. Hmmmmmm.

**Comment by justvanrossum** [comment]
You mean, to maintain the selection ofter sorting? Or what is the problem?

**Comment by typesupply** [comment]
I was looking into your suggestion of storing the arranged indexes instead of the arranged items. When I call sorted on duplicate items, I don't know how I can work out what the unsorted index is of the sorted item. Better example:

```python
import operator

items = [
    dict(value="A"),
    dict(value="A"),
    dict(value="B"),
    dict(value="B"),
    dict(value="C"),
    dict(value="D"),
]

arrangedIndexes = []
for item in sorted(items, key=operator.itemgetter("value")):
    index = ?
    arrangedIndexes.append(index)
```

**Comment by justvanrossum** [comment]
It can be done like this (leaving out the operator generalization):

```python
def sortedIndices(items):
    indices = list(range(len(items)))
    arrangedIndexes = []
    def keyFunc(index):
        return items[index]["value"]
    return sorted(indices, key=keyFunc)

items = [
    dict(value="A"),
    dict(value="A"),
    dict(value="B"),
    dict(value="B"),
    dict(value="C"),
    dict(value="D"),
]

print(sortedIndices(items))
```

**Comment by typesupply** [comment]
@typemytype

Are both `menuCallback` and `setMenu` needed? I can see a theoretical need for both, but want to make sure I'm not adding anything unnecessary.

---

@justvanrossum 

I've implemented everything with indexes internally. Thanks for the advice on all of that. I was looking through the old List code and saw how I handled items that compare the same before. In short, I made a mapping each time (!!!!!!!!!!) I needed to find an index when sort descriptors were present:

```python
sortedObjects = [(id(sortedArray[index]), sortedArray[index]) for index in indexes]
sortedObjects = dict.fromkeys(sortedObjects)
```

This retroactively makes me shudder, but it's been there since a CVS commit on May 29, 2006. Seeing this made me wonder if I could do something similar in `setSelectedItems` to eliminate the compare alike problem:

```python
    def setSelectedItems(self, items):
        allItems = self.get()
        indexMapping = {
            id(item) : index
            for index, item in enumerate(allItems)
        }
        selectionIndexes = [
            indexMapping[id(item)]
            for item in items
        ]
        self.setSelectedIndexes(selectionIndexes)
```

I have an aversion to using `id` for anything other than debugging these days, but maybe this would be okay in this controlled situation?

**Comment by justvanrossum** [comment]
Hmhm. I think it nicely demonstrates what a disaster `setSelectedItems()` is... I don't know anything better.

All I can do is point out another flaw -- it will surprise people who try this:

```python
        self.w.listThing.setSelectedItems([dict(a=12, b=13)])
```
But perhaps it's ok, if the documentation says that this will compare by object id, not equality.

**Comment by typesupply** [comment]
I've got the basics complete and ready for testing if anyone wants to give it a try. It's in the list2 branch. There are a couple of tests in tests/list2.

I'll keep working through the minor stuff on the checklist above while I work up the energy to deal with drag and drop.

**Comment by typesupply** [comment]
FYI, we have fully customizable group rows now:

<img width="412" alt="image" src="https://user-images.githubusercontent.com/4992934/115147110-6692d680-a027-11eb-991b-be9eff5569e4.png">

**Comment by typesupply** [comment]
I'm starting to get a new drag and drop API in place for List2 and any other vanilla view that would want it (I'm thinking Group and Box). Along the way, I've had the need to define methods that are used across > 1 NS* subclasses. Mixins aren't possible with PyObjC, so I've been trying to figure out a way to write this code once and share it across the different classes. Is the following code a terrible idea?

I define the methods as functions once along with a function that will add the methods to a class:

```python
def performDragOperation_(self, draggingInfo):
    # blah

def assignDropProtocol(cls):
    cls.performDragOperation_ = performDragOperation_
```

In the various vanilla files, I import the application function and use it:

```python
class VanillaGroupViewSubclass(NSView): pass

assignDropProtocol(VanillaGroupViewSubclass)

class Group(VanillaBaseObject):

    nsViewClass = VanillaGroupViewSubclass


class VanillaBoxSubclass(NSBox): pass

    nsBoxClass = VanillaBoxSubclass

assignDropProtocol(VanillaBoxSubclass)

class Box(VanillaBaseObject):

    nsBoxClass = VanillaBoxViewSubclass


class VanillaTableViewSubclass(NSTableView): pass

assignDropProtocol(VanillaTableViewSubclass)

class List2(ScrollView):

    nsTableViewClass = VanillaList2TableViewSubclass
```

This works in my limited testing, but I feel like I'm doing something that I shouldn't. However, I haven't been able to come up with another solution...

**Comment by justvanrossum** [comment]
> Mixins aren't possible with PyObjC

Syntactically it works, but the methods from the mixin don't become objc selectors.

It seems that injecting a method after the class was built also doesn't quite work: I don't think `performDragOperation_` becomes a "proper" objc selector.

It can be made to work with a metaclass (adding to ClassNameIncrementer because it was there) :

```python
import AppKit

def ClassNameIncrementer(clsName, bases, dct):
    import objc
    orgName = clsName
    counter = 0
    while True:
        try:
            objc.lookUpClass(clsName)
        except objc.nosuchclass_error:
            break
        counter += 1
        clsName = orgName + str(counter)
    for base in bases[1:]:
        for n in dir(base):
            if n[0] != "_":
                if n not in dct:
                    dct[n] = getattr(base, n)
    bases = bases[:1]
    return type(clsName, bases, dct)


class Mixin:
    def bar_(self, arg):
        print("bar", arg)

class Test(AppKit.NSObject, Mixin, metaclass=ClassNameIncrementer):
    def foo_(self, arg):
        print("foo", arg)

t = Test.alloc().init()
t.foo_(124)
t.bar_(345)
print(t.foo_)
print(t.bar_)
```

That may be a bit of a hack, though.

How about those classes all do individually implement the needed method, but all they do is call a common Python function?

**Comment by typesupply** [comment]
> How about those classes all do individually implement the needed method, but all they do is call a common Python function?

Yeah, they call a mixin on the vanilla wrapper. Like this:

```python
def draggingEntered_(self, draggingInfo):
    return self.vanillaWrapper().dropCandidateEntered(draggingInfo)

def draggingUpdated_(self, draggingInfo):
    return self.vanillaWrapper().dropCandidateUpdated(draggingInfo)
```

There's no interpretation done in the NS methods, it routes it and returns the result.

I could copy and paste the code as needed.

**Comment by justvanrossum** [comment]
> I could copy and paste the code as needed.

Why not. Sounds the cure for avoiding copy/paste may be worse than the disease...

**Comment by typesupply** [comment]
> Why not. Sounds the cure for avoiding copy/paste may be worse than the disease...

Okay. Thanks!

Getting there:

<img width="620" alt="IMG_7558" src="https://user-images.githubusercontent.com/4992934/117147520-54cd6500-ad83-11eb-85e8-0a05e40d62d7.png">

**Comment by typesupply** [comment]
I'm working on reordering items within a list which raises a logic puzzle to solve. ([Example of someone else trying to solve this.](https://www.natethompson.io/2019/03/23/nstableview-drag-and-drop.html#move-rows) This won't work for us because vanilla doesn't expose the `tableView.moveRowAtIndex` method.) Off the top of your head, does anyone have an algorithm for reordering items in a list robustly? I can figure one out, but I figured I'd ask before I sidetrack myself from working out all the AppKit stuff. Edge cases: multiple items selected, items moving to position after their current position, etc.

**Comment by typesupply** [comment]
Okay. List2 is ready for testing and potential use by someone other than me. Here are tests (located in lib/vanilla/tests):


- [List2 Cells](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/test/list2/cells.py)
- [List2 Selection](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/test/list2/selection.py)
- [List2 Group Rows](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/test/list2/groupRows.py)
- [Drag and Drop](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/test/testDragAndDrop.py)

The drag and drop API is documented here (it needs to be linked in the vanilla doc ReST):

- [List2](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/vanillaList2.py)
- [Drag and Drop](https://github.com/robotools/vanilla/blob/list2/Lib/vanilla/dragAndDrop.py)
- Group has a new `dropSettings` arg in `__init__` but it's the `dropSettings` documented in the d&d documentation.

**Comment by typesupply** [comment]
@typemytype and @benkiel: We merged the list2 branch into the main branch, right? Would it be safe for me to delete that branch and close this issue?

**Comment by benkiel** [comment]
Yes, here: https://github.com/robotools/vanilla/pull/153
I've deleted the branch, will close issue.
